ops/_tracing/buffer.py

```diff
@@ -0,0 +1,244 @@
+    # when database is written to from two threads,
+    # either we'd get an exaception on BEGIN (immediate mode)
+    # or we'd get an exception on COMMIT (deferred mode)
+    # either way, a retry will be needed sooner or later.
+
+    @retry
+    def mark_observed(self):
+        if self.observed:
+            return
+
+        with self.tx(timeout=LONG_DB_TIMEOUT) as conn:
+            conn.execute(
+                f"""
+                UPDATE tracing
+                SET priority = ?
+                WHERE id IN ({','.join(('?',) * len(self.ids))})
+                """,  # noqa: S608
+                (OBSERVED_PRIORITY, *tuple(self.ids)),
+            )
+        self.observed = True
+        self.ids.clear()
+
+    @retry
+    def pump(self, chunk: bytes | None = None) -> tuple[int, bytes] | None:
+        """Pump the buffer queue.
+
+        Accepts an optional new data chunk.
+        Removes old, boring data if needed.
+        Returns the oldest important record.
+        """
+        # NOTE: discussion about transaction type:
+        # - this may be a read-only transaction (no data to save, read out one record)
+        # - or a read transaction later upgraded to write (check space, then delete some)
+        # currently I've made `self.tx()` return a write transaction always
+        # which is safer, but may incur a filesystem modification cost.
+        with self.tx(readonly=not chunk) as conn:
+            if chunk:
+                # Ensure that there's enough space in the buffer
+                chunklen = (len(chunk) + 4095) // 4096 * 4096
+                stored: int | None = conn.execute(
+                    """
+                    SELECT sum((length(data)+4095)/4096*4096)
+                    FROM tracing
+                    """
+                ).fetchone()[0]
+                # TODO: expose `stored` in metrics, one day
+                excess = (stored or 0) + chunklen - BUFFER_SIZE
+                logging.debug(f'{excess=}')
+
+                if excess > 0:
+                    # Drop lower-priority, older data
+                    cursor = conn.execute(
+                        """
+                        SELECT id, (length(data)+4095)/4096*4096
+                        FROM tracing
+                        ORDER BY priority ASC, id ASC
+                        """
+                    )
+
+                    collected_ids: set[int] = set()
+                    collected_size: int = 0
+                    for id_, size in cursor:
+                        collected_ids.add(id_)
+                        collected_size += size
+                        if collected_size > excess:
+                            break
+
+                    assert collected_ids
+                    logging.debug(f'{len(collected_ids)=}')
+                    conn.execute(
+                        f"""
+                        DELETE FROM tracing
+                        WHERE id IN ({','.join(('?',) * len(collected_ids))})
```

Comment:
I think it'd avoid allocating/copying twice to say `'?,' * (len(collected_ids)-1) + '?'`

Reply:
That would not be the same if the collected_ids set is empty.